src/blocked.c

```diff
@@ -863,10 +466,196 @@ static void signalKeyAsReadyLogic(redisDb *db, robj *key, int type, int deleted)
+             * 2. We want to serve clients blocked on module keys
+             *    regardless of the object type: we don't know what the
+             *    module is trying to accomplish right now.
+             * 3. In case of XREADGROUP call we will want to unblock on any change in object type
+             *    or in case the key was deleted, since the group is no longer valid. */
+            if ((o != NULL && (receiver->bstate.btype == getBlockedTypeByType(o->type))) ||
+                (o != NULL && (receiver->bstate.btype == BLOCKED_MODULE)) ||
+                (receiver->bstate.unblock_on_nokey))
+            {
+                if (receiver->bstate.btype != BLOCKED_MODULE)
+                    unblockClientOnKey(receiver, rl->key);
+                else
+                    moduleUnblockClientOnKey(receiver, rl->key);
+            }
+        }
+    }
+}
+
+/* block a client due to wait command */
+void blockForReplication(client *c, mstime_t timeout, long long offset, long numreplicas) {
+    c->bstate.timeout = timeout;
+    c->bstate.reploffset = offset;
+    c->bstate.numreplicas = numreplicas;
+    listAddNodeHead(server.clients_waiting_acks,c);
+    blockClient(c,BLOCKED_WAIT);
+}
+
+/* Postpone client from executing a command. For example the server might be busy
+ * requesting to avoid processing clients commands which will be processed later
+ * when the it is ready to accept them. */
+void blockPostponeClient(client *c) {
+    c->bstate.timeout = 0;
+    blockClient(c,BLOCKED_POSTPONE);
+    listAddNodeTail(server.postponed_clients, c);
+    c->postponed_list_node = listLast(server.postponed_clients);
+    /* Mark this client to execute its command */
+    c->flags |= CLIENT_PENDING_COMMAND;
+}
+
+/* Block client due to shutdown command */
+void blockClientShutdown(client *c) {
+    blockClient(c, BLOCKED_SHUTDOWN);
+    /* Mark this client to execute its command */
+}
+
+/* Unblock a client once a specific key became available for it.
+ * This function will remove the client from the list of clients blocked on this key
+ * and also remove the key from the dictionary of keys this client is blocked on.
+ * in case the client has a command pending it will process it immediately.  */
+static void unblockClientOnKey(client *c, robj *key) {
+    dictEntry *de;
+
+    de = dictFind(c->bstate.keys, key);
+    releaseBlockedEntry(c, de, 1);
+
+    /* Only in case of blocking API calls, we might be blocked on several keys.
+       however we should force unblock the entire blocking keys */
+    serverAssert(c->bstate.btype == BLOCKED_STREAM ||
+                c->bstate.btype == BLOCKED_LIST   ||
+                c->bstate.btype == BLOCKED_ZSET);
+
+    unblockClient(c);
+    /* In case this client was blocked on keys during command
+     * we need to
+     */
+    if (c->flags & CLIENT_PENDING_COMMAND) {
+        c->flags &= ~CLIENT_PENDING_COMMAND;
+        processCommandAndResetClient(c);
```

Comment:
or add a new flag to make the unblocked client call `call()` directly

Reply:
@soloestoy not sure i understand the context.
is this meant to be related to this discussion:
https://github.com/redis/redis/pull/11012#discussion_r984684026

@ranshid isn't that discussion resolved?
i don't know what was in the code when the comment was placed, but it looks to me as if there's no recursion now.

Reply:
@oranagra until now I did not understand @madolson, but I think that indeed there is an issue here:
lets say I have 4 clients (a,b,c) each blocked on lists (l1,l2,l3) respectfully and in that order.
now I have client H calling `MULTI; LPUSH l1 1; LPUSH l2 2 ; EXEC`
So client H will  also call handleClientsBlockedOnKeys which will take the current list of ready_keys aside and iterate over it.
the first key it will go over is l1 for which the only blocked client in the list will be client 'a' which will be unblocked and will call processCommandAndResetClient.
But lets say client 'a' is doing a `BLPOPRPUSH l1 l3 0` which will signal l3 as ready?
So it would have called handleClientsBlockedOnKeys (recursive call), which will signal l3 as ready and then it would process client c before processing client b, which was blocked first.

I am not sure how big of a problem this is (it cannot happen in case l2 == l3) but this gives me motivation to follow @soloestoy suggestion

Reply:
ohh, now i see what i was missing, handleClientsBlockedOnKeys is not called only from beforeSleep, but also from processCommand.

i'm not sure it matters that it'll process client c before client b, but i do fear the recursion.
maybe we can test some flag so that processCommand doesn't do that, and then the loop in handleClientsBlockedOnKeys will take care of the rest?

Reply:
Actually I do think there is an issue with that:
For example what if client 'c' was unblocked first and running `BLMPOP l2 l3` - so it would get precedence over client 'b' on popping from list l2... which is a BUG IMO.

Reply:
i don't think that's a bug.. or maybe i'm missing something.
are you saying that because you know the order in which the commands arrived to redis?
from the client's perspective they might have arrived in the order [a,c,b] and letting c pop something that was pushed by a as well as something that was pushed by whoever released a from blocking.

Reply:
I agree that in case the 2 clients weere blocked together the user might not notic.
But our documentation clearly states:
```
If multiple clients are blocked for the same key, the first client to be served is the one that was waiting for more time (the first that blocked for the key). 
```
and it can be that client 'c' was blocked way longer after client 'b'

Reply:
ok, you're right. it does violate that statement (and fairness).
we can add a test for it...

so in theory, we have the server.ready_keys list that should keep the order, but because handleClientsBlockedOnKeys messes with this list (replaces it), and because of the recursion, we can run into that situation, right?

i do think we wanna eliminate the recursion anyway.
so maybe setting a flag so that processCommand doesn't call handleClientsBlockedOnKeys in that flow?

Reply:
@oranagra I have introduced a flag to prevent the recursion. I could use the new CLIENT_REPROCESSING_COMMAND flag. but I did not want to hard link the logic of reprocessing with the fact that we are doing handleClientsBlockedOnKeys

Reply:
done